### PR TITLE
gh-109802: Increase test coverage for complexobject.c

### PR DESCRIPTION
**Pull request by skirpichev**
* _Py_c_pow: L139, L135-136
* complex_hash: L423
* to_complex: L443, L449, L453-454
* complex_sub: L474, L475
* complex_mul: L485, L486
* complex_div: L496, L497
* complex_pow: L512, L513, L522
* complex_pos: L555-L559
* complex_richcompare: L616, L625
* complex_from_string_inner: L786, L804
* complex_new_impl: L944, L951, L958, L965
* complex_bool: L580
* tests for _Py_c_* C API

Line numbers are wrt to [complexobject.c at 54fbfa8d5e](https://github.com/python/cpython/blob/54fbfa8d5e52ed315d10640c875a8b8dd2c30cec/Objects/complexobject.c).

----------

This is #109642 without changes in complexobject.c (some dead code will be uncovered by tests in test_complex.py, test_cmath.py and test_capi/test_complex.py).

* Issue: gh-109802

### CODE REVIEW

**Comment by skirpichev** [comment]
CC @serhiy-storchaka this is #109642 without removal of dead code.  (tests only)  Description was updated to mention covered lines.

File: Modules/_testcapi/complex.c [hunk 1]
```diff
@@ -85,6 +87,83 @@ complex_asccomplex(PyObject *Py_UNUSED(module), PyObject *obj)
...
+        if (!res) {                                              \
+            return NULL;                                         \
+        }                                                        \
+        err = PyLong_FromLong(errno);                            \
+        if (!err) {                                              \
+            Py_DECREF(res);                                      \
+            return NULL;                                         \
+        }                                                        \
+                                                                 \
+        return Py_BuildValue("NN", res, err);                    \
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/complex.c]: Why not use simply `Py_BuildValue("di", num, errno)`?

File: Modules/_testcapi/complex.c [hunk 2]
```diff
@@ -85,6 +87,83 @@ complex_asccomplex(PyObject *Py_UNUSED(module), PyObject *obj)
...
+        Py_complex num, exp;                                     \
+        PyObject *res, *err;                                     \
+                                                                 \
+        if (!PyArg_ParseTuple(args, "DD", &num, &exp)) {         \
+            return NULL;                                         \
+        }                                                        \
+                                                                 \
+        errno = 0;                                               \
+        num = _Py_c_##suffix(num, exp);                          \
+        _Py_ADJUST_ERANGE2(num.real, num.imag);                  \
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/complex.c]: Why call `_Py_ADJUST_ERANGE2`? Aren't these tests for `_Py_c_XXX` functions?

File: Modules/_testcapi/complex.c [hunk 3]
```diff
@@ -85,6 +85,65 @@ complex_asccomplex(PyObject *Py_UNUSED(module), PyObject *obj)
...
+                                                                 \
+        errno = 0;                                               \
+        num = _Py_c_##suffix(num, exp);                          \
+                                                                 \
+        res = PyComplex_FromCComplex(num);                       \
+        if (!res) {                                              \
+            return NULL;                                         \
+        }                                                        \
+                                                                 \
+        return Py_BuildValue("Ni", res, errno);                  \
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/complex.c]: Why not use "D"?

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @skirpichev.

**Pull request merged**